### PR TITLE
25 binop function

### LINKED ISSUE #25: Attach binop function to AST
# Description

Some binary operators need a special function to get evaluated. This will need to get attached to the AST.

# Sub-tasks

- [x] Make binop use a special function
- [x] Make annotateAst attach the function to the ast

### PR DESCRIPTION
**Pull request by HectorPeeters**
# Description

This pull request implements the new approach for annotating the binary operator

Closes #25 

# Changes

1. Binary operator annotation now attaches a function 

# Deletes Source Branch?

Yes:    Work is done, the branch can be safely deleted.

# How Many Approvals?

* Approval of all (Controversial changes that change the entire project)

### CODE REVIEW

**Review by FBroy — APPROVED** [review]
This is a pretty nice solution for the problem that we were facing!

File: src/typeInference/annotateAST.ts [hunk 1]
```diff
@@ -34,43 +34,43 @@ export function annotate(ast: IAST): SequenceType | undefined {
...
 			};
 
-			insertAttribute(ast, stringSequenceType);
+			insertAttribute(ast, 'type', stringSequenceType);
 			return stringSequenceType;
 		default:
 			for (let i = 1; i < ast.length; i++) {
 				annotate(ast[i] as IAST);
 			}
-			return { type: ValueType.XSERROR, mult: SequenceMultiplicity.EXACTLY_ONE };
```
Comment (yuanzexiong) [inline src/typeInference/annotateAST.ts:74]: Why is the error removed btw?

**Review by maanpalle — APPROVED** [review]
Looks nice good job

File: src/typeInference/annotateBinaryOperator.ts [hunk 2]
```diff
@@ -1,165 +1,108 @@
...
 import { insertAttribute } from './insertAttribute';
 
-type BinOpLookupTable = {
-	[key: number]: ValueType;
+function hash(left: ValueType, right: ValueType, op: string): number {
+	return (
+		((left as number) << 20) +
+		((right as number) << 12) +
+		(op.charCodeAt(0) << 8) +
+		op.charCodeAt(1)
```
Comment (yuanzexiong) [inline src/typeInference/annotateBinaryOperator.ts:14]: Shifting is a cool idea, if typescript manages to keep up the performance and does not do some weird things

File: src/typeInference/annotateAST.ts (hunk 1 above)
Comment (HectorPeeters) [reply src/typeInference/annotateAST.ts:74]: It's not really an error, it something I added at the start of the prototype to mark that something could not get annotated. However this changed to just returning undefined later and this snippet of code was still left in there.

**Review by MTimmerije — APPROVED** [review]
Looked through it all, I like the solution for the function calls.

File: src/typeInference/annotateBinaryOperator.ts [hunk 3]
```diff
@@ -1,165 +1,108 @@
...
-		isSubtypeOf(right.type, ValueType.XSNUMERIC)
-	) {
-		const type = {
-			type: BINOP_LOOKUP[left.type + right.type * 1000] || ValueType.XSDECIMAL,
-			mult: left.mult,
-		};
-		insertAttribute(ast, type);
+	const funcData = BINOP_EVAL_FUNCTIONS[hash(left.type, right.type, operator)];
+	if (funcData) {
+		const type = { type: funcData[1], mult: left.mult };
```
Comment (yuanzexiong) [inline src/typeInference/annotateBinaryOperator.ts:99]: Probably better to add a comments or doc on what is funcData[0] and [1]

**Review by yuanzexiong — APPROVED** [review]
Seems pretty good and btw we need to start documenting the code a bit otherwise it might get out of hands

File: src/typeInference/annotateBinaryOperator.ts (hunk 2 above)
Comment (HectorPeeters) [reply src/typeInference/annotateBinaryOperator.ts:14]: I think it will work for now but this might have to be changed in the future.

**Pull request merged**